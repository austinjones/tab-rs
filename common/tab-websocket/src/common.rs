use crate::WebSocket;
use futures::SinkExt;
use log::error;
use serde::Serialize;
use tungstenite::{Error, Message};

pub async fn send_message<Msg: Serialize>(websocket: &mut WebSocket, message: Msg) {
    let encoded = bincode::serialize(&message);

    if let Err(e) = encoded {
        error!("failed to encode message: {}", e);
        return;
    }

    let sent = websocket.send(Message::Binary(encoded.unwrap())).await;
    if let Err(e) = sent {
        error!("failed to send message: {}", e);
    }
}

pub async fn send_close(websocket: &mut WebSocket) {
    if let Err(e) = websocket.send(Message::Close(None)).await {
        match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                return;
            }
            _ => {
                error!("failed to send close frame: {}", e);
            }
        }
    }
}
