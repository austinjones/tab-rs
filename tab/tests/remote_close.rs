use insta::assert_snapshot;

mod common;
use common::*;

/// Tests that two sessions can be established (A and B),
///  that A can be reconnected to,
///  and B can be switched to by executing the tab binary within A.
#[tokio::test]
async fn remote_close() -> anyhow::Result<()> {
    let session = TestSession::new()?;

    remote_close_env(&session).await?;

    remote_close_name(&session).await?;

    Ok(())
}

async fn remote_close_env(session: &TestSession) -> anyhow::Result<()> {
    let tab = format!("remote_close_env/");

    let result = session
        .command()
        .tab(tab.as_str())
        .await_stdout("$", 3000)
        .stdin("$TAB_BIN --close\n")
        .await_stdout("close", 300)
        .complete_snapshot()
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("env", result.snapshot);

    Ok(())
}

async fn remote_close_name(session: &TestSession) -> anyhow::Result<()> {
    let tab = format!("remote_close_name/");

    let result = session
        .command()
        .tab(tab.as_str())
        .await_stdout("$", 3000)
        .stdin(format!("TAB_ID='' $TAB_BIN --close {}\n", tab))
        .await_stdout("close remote_close_name/", 300)
        .complete_snapshot()
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("name", result.snapshot);

    Ok(())
}
