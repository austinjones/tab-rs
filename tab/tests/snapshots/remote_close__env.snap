---
source: tab/tests/remote_close.rs
expression: result.snapshot
---
$ $TAB_BIN --close
